weather_mv/loader_pipeline/ee_util.py

```diff
@@ -0,0 +1,457 @@
+from urllib.parse import urlparse
+import ee
+import xarray as xr
+import subprocess
+from retry import retry
+import re
+
+import apache_beam as beam
+from apache_beam.options.pipeline_options import PipelineOptions
+from google.api_core.exceptions import BadRequest
+from google.api_core.exceptions import NotFound
+from google.cloud import storage
+from google.auth import compute_engine, default
+from google.auth.transport import requests
+
+from .sinks import ToDataSink, open_dataset, open_local
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+CANARY_BUCKET_NAME = 'anthromet_canary_bucket'
+CANARY_RECORD = {'foo': 'bar'}
+CANARY_RECORD_FILE_NAME = 'canary_record.json'
+
+
+def is_compute_engine():
+    """
+    Determines if the application in running in Compute Engine Environment
+
+    Returns:
+    a bool indicating whether the application is running in compute engine
+    """
+    COMPUTE_ENGINE_STR = 'Metadata-Flavor: Google'
+    command = ['curl', 'metadata.google.internal', '-i']
+    result = subprocess.run(command, stdout=subprocess.PIPE)
+    result_output = result.stdout.decode('utf-8')
+    return COMPUTE_ENGINE_STR in result_output
+
+
+# Concurrent calls to OAuth can cause problems. See yaqs/8220221608137064448.
+def get_creds(use_personal_account):
+    """
+    Fetches credentials for authentication
+
+    Note: If the application is running in compute engine, it will use credentials
+    of service account bound to the VM otherwise will try to use user credentials.
+
+    Returns:
+        credentials
+    """
+    # Authenticate with pop-up if credentials are not available
+    if use_personal_account:
+        ee.Authenticate()
+
+    if is_compute_engine() and not use_personal_account:
+        # To use the default service account, it is required to register the service account
+        # with EE from here: https://signup.earthengine.google.com/#!/service_accounts
+        # See https://developers.google.com/earth-engine/guides/service_account#register-the-
+        # service-account-to-use-earth-engine
+        # for more detail.
+        creds = compute_engine.Credentials()
+    else:
+        creds, _ = default()
+    request = requests.Request()
+    creds.refresh(request)
+    return creds
+
+
+def ee_initialize(use_personal_account=False,
+                  enforce_high_volume=False,
+                  allow_old_versions=False):
+    """Initializes ee with the high volume API when using a compute engine VM.
+
+    Once earth engine is initialized, repeated calls will still verify the high
+    volume api and old version constrains, but will then no-op.
+
+    Raises:
+        RuntimeError: Earth Engine did not initialize.
+    """
+    creds = get_creds(use_personal_account)
+
+    # Only the compute engine service service account can access the high volume api.
+    # The high volume api is only available with ee version 0.1.263 or later.
+
+    up_to_date = ee.__version__ >= '0.1.263'
```

Comment:
Do we want to support all versions of EE? Why not set the minimum version of which ee API we use?